src/components/tabs/ScatterPlotTab.tsx

```diff
@@ -386,6 +392,131 @@ export default memo(function ScatterPlotTab(props: ScatterPlotTabProps): ReactEl
     );
   };
 
+  /**
+   * Applies coloring to point traces in a scatterplot. Does this by splitting the data into multiple traces each with a solid
+   * color, which is much faster than using Plotly's native color ramping. Also enforces a maximum number of points
+   * per trace, which significantly speeds up Plotly renders.
+   *
+   * @param xData
+   * @param yData
+   * @param objectIds
+   * @param trackIds
+   * @param markerConfig Additional marker configuration to apply to all points. By default,
+   * markers are size 4.
+   * @param {Color | undefined} overrideColor When defined, uses a base color for all points, instead of
+   * calculating based on the color ramp or palette.
+   */
+  const colorizeScatterplotPoints = (
+    xData: DataArray,
+    yData: DataArray,
+    objectIds: number[],
+    trackIds: number[],
+    markerConfig: Partial<PlotMarker> = {},
+    overrideColor?: Color
+  ): Partial<PlotData>[] => {
+    if (selectedFeatureName === null || dataset === null || !xAxisFeatureName || !yAxisFeatureName) {
+      return [];
+    }
+
+    const featureData = dataset.getFeatureData(selectedFeatureName);
+
+    const traceDataBuckets: TraceData[] = [];
+    if (!featureData || markerConfig.color || overrideColor) {
+      // Do no coloring! Keep all points in the same bucket.
+      traceDataBuckets.push({
+        x: Array.from(xData),
+        y: Array.from(yData),
+        objectIds: objectIds,
+        trackIds: trackIds,
+        color: overrideColor || new Color(),
+      });
+    } else {
+      // Generate colors
+      const categories = dataset.getFeatureCategories(selectedFeatureName);
+      const isCategorical = categories !== null;
+      const colors = isCategorical
+        ? categoricalPalette.slice(0, categories.length)
+        : subsampleColorRamp(colorRamp, COLOR_RAMP_SUBSAMPLES);
+      const colorMinValue = isCategorical ? 0 : colorRampMin;
+      const colorMaxValue = isCategorical ? categories.length - 1 : colorRampMax;
+
+      // Make a bucket group for each ramp/palette color and for the out-of-range and outliers.
+      // index 0 = out of filter range, index 1 = outliers.
+      for (let i = 0; i < colors.length + 2; i++) {
+        let color;
+        if (i === 0) {
+          color = viewerConfig.outOfRangeDrawSettings.color;
+        } else if (i === 1) {
+          color = viewerConfig.outlierDrawSettings.color;
+        } else {
```

Comment:
Believe I commented this in a child PR (whoops), but could you move these into two initial `push` calls and get rid of the `if` inside this loop?

Reply:
Yup, should be fixed in the child PR!